### PR TITLE
Remove default token volume and volume_mount from Pod state

### PR DESCRIPTION
**Pull request by dak1n1**
Fixes https://github.com/hashicorp/terraform-provider-kubernetes/issues/1085

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [ ] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
```

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-kubernetes/blob/master/CHANGELOG.md):

```release-note
Ensure default service account token is not saved in Pod state.
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

File: kubernetes/structures_pod.go [hunk 1]
```diff
@@ -107,6 +108,23 @@ func flattenPodSpec(in v1.PodSpec) ([]interface{}, error) {
...
+		if in.AutomountServiceAccountToken != nil && *in.AutomountServiceAccountToken == true {
+			for i, volume := range in.Volumes {
+				nameMatchesDefaultToken, err := regexp.MatchString("default-token-([a-z0-9]{5})", volume.Name)
+				if err != nil {
+					return []interface{}{att}, err
+				}
+				if nameMatchesDefaultToken && volume.Secret != nil {
+					for _, secretVolume := range volume.Secret.Items {
+						if secretVolume.Path == "/var/run/secrets/kubernetes.io/serviceaccount" {
+							removeVolume(i, in.Volumes)
```
Comment (jrhouston) [inline kubernetes/structures_pod.go:122]: The reason this isn't working is because your `removeVolume` func returns a new slice, so you need to assign that new slide to `in.Volumes`.

File: kubernetes/structures_pod.go [hunk 2]
```diff
@@ -107,6 +108,23 @@ func flattenPodSpec(in v1.PodSpec) ([]interface{}, error) {
 	}
 
 	if len(in.Volumes) > 0 {
+		// If the serviceAccountToken is being added to the list of volumes,
+		// we need to remove it prior to recording the state.
+		if in.AutomountServiceAccountToken != nil && *in.AutomountServiceAccountToken == true {
+			for i, volume := range in.Volumes {
+				nameMatchesDefaultToken, err := regexp.MatchString("default-token-([a-z0-9]{5})", volume.Name)
```
Comment (jrhouston) [inline kubernetes/structures_pod.go:115]: What I would do is put this logic [here](https://github.com/hashicorp/terraform-provider-kubernetes/blob/master/kubernetes/structures_pod.go#L256) in the loop that expands the volumes and use a `continue` statement to skip it, rather than removing it here.

**Comment by dak1n1** [comment]
Hmm, well this does look neater. And it successfully skips adding both the volume and volumeMount to state... well, almost. It adds them as empty blocks if I put the logic into `flattenVolumes()` and `flattenContainerVolumeMounts()`, rather than in `flattenPodSpec()` . ("Empty blocks" meaning: `volume{}` and `volume_mount{}`). Here is the state for context:

https://gist.githubusercontent.com/dak1n1/3ec5fc36cb9c4ab16d7d5ed8c159009c/raw/bc670e3486d17d5eb7eafe8016ebb1e63d09d06f/gistfile1.txt

However, if I remove these at the PodSpec level, it omits the blocks entirely, which prevents the diff. I'm going to explore that route next.

**Comment by dak1n1** [comment]
Got it working! Pods are no longer destroyed on each apply.

```
[dakini@dax pod_with_volume]$ terraform show 
[...]
            volume_mount {
                mount_path        = "/etc/test"
                mount_propagation = "None"
                name              = "pvc"
                read_only         = false
            }
        }

        volume {
            name = "pvc"

            persistent_volume_claim {
                claim_name = "test"
                read_only  = false
            }
        }
    }
}
[dakini@dax pod_with_volume]$ terraform plan
kubernetes_persistent_volume_claim.test: Refreshing state... [id=default/test]
kubernetes_pod.main: Refreshing state... [id=default/test]

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.
```

File: kubernetes/structures_pod.go [hunk 3]
```diff
@@ -107,6 +108,17 @@ func flattenPodSpec(in v1.PodSpec) ([]interface{}, error) {
 	}
 
 	if len(in.Volumes) > 0 {
+		for i, volume := range in.Volumes {
+			// To avoid perpetual diff, remove the default service account token volume from PodSpec.
+			nameMatchesDefaultToken, err := regexp.MatchString("default-token-([a-z0-9]{5})", volume.Name)
```
Comment (jrhouston) [inline kubernetes/structures_pod.go:113]: Ah, I just realized we are going to have to be a bit more clever than this. 

When the token get's auto-mounted it takes the name from the service account, so only if they don't supply a service account will it be called `default`, otherwise it will be the name of the service account. You can see how it does this in the admission controller code for it: [kubernetes/blob/master/plugin/pkg/admission/serviceaccount/admission.go#L46](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/serviceaccount/admission.go#L468)

So we'll have to actually pass down the serviceAccount name and make the regex match that. 😢 

We could just match on `-token-` but that creates a surprising edge case for someone to discover.

Comment (dak1n1) [reply kubernetes/structures_pod.go:113]: oh, nice! That is really helpful. I hadn't thought to test with a service account specified. I'll put in a test to catch these cases and make it clever enough to handle them. Thanks!

**Comment by dak1n1** [comment]
It works now. I'll put the test into PR #1074, since that contains the test for bug #1085

File: kubernetes/structures_pod.go [hunk 4]
```diff
@@ -107,6 +116,17 @@ func flattenPodSpec(in v1.PodSpec) ([]interface{}, error) {
...
 
 	if len(in.Volumes) > 0 {
+		for i, volume := range in.Volumes {
+			// To avoid perpetual diff, remove the service account token volume from PodSpec.
+			nameMatchesDefaultToken, err := regexp.MatchString(serviceAccountRegex, volume.Name)
+			if err != nil {
+				return []interface{}{att}, err
+			}
+			if nameMatchesDefaultToken {
+				in.Volumes = removeVolumeFromPodSpec(i, in.Volumes)
```
Comment (jrhouston) [inline kubernetes/structures_pod.go:126]: At this point you can `break` the loop because you've found the index of the volume and so don't need to keep looping.

File: kubernetes/structures_container.go [hunk 5]
```diff
@@ -404,6 +406,17 @@ func flattenContainers(in []v1.Container) ([]interface{}, error) {
...
 
 		if len(v.VolumeMounts) > 0 {
+			for num, m := range v.VolumeMounts {
+				// To avoid perpetual diff, remove the default service account token volume from the container's list of volumeMounts.
+				nameMatchesDefaultToken, err := regexp.MatchString(serviceAccountRegex, m.Name)
+				if err != nil {
+					return att, err
+				}
+				if nameMatchesDefaultToken {
+					v.VolumeMounts = removeVolumeMountFromContainer(num, v.VolumeMounts)
```
Comment (jrhouston) [inline kubernetes/structures_container.go:416]: Can also just `break` here.

**Review by jrhouston — CHANGES_REQUESTED** [review]
Looks good just a couple of nits.

**Review by jrhouston — APPROVED** [review]
🚀
